### DISCUSSION TITLE
Installation folder accoring to property defined in command line

### DISCUSSION THREAD

**Discussion by idaamit** [wixtoolset/issues#7186] [Questions] [answered]
Currently, my product is installed on pre-defined ProgramFilesFolder property :  		
```
<Directory Id="TARGETDIR" Name="SourceDir">
        <Directory Id="ProgramFilesFolder">
	      <Directory Id="INSTALLFOLDER" Name="BuiltInPropertiesExample" />
	</Directory>
</Directory>
```
I would like to install it under a different location according to property provided in the command line. for example : 
`msiexec /i BuiltInPropertiesExample.msi  **LOCATION=D:\MyProduct\BuiltInPropertiesExample** `
but if LOCATION property is not provided in the command, I would like to keep the current behavior: **C:\Program Files\BuiltInPropertiesExample**
How do I make such installation ?

**Comment by fyodorkor** [comment, marked as answer]
You add a CustomAction of type SetDirectory after the CostFinalize standard
action that sets INSTALLFOLDER to your property. Add a condition to this
action to do it only if your property is populated.

On Tue, Jan 31, 2023 at 12:20 PM idaamit ***@***.***> wrote:

> Currently, my product is installed on pre-defined ProgramFilesFolder
> property :
>
> <Directory Id="TARGETDIR" Name="SourceDir">
>         <Directory Id="ProgramFilesFolder">
> 	      <Directory Id="INSTALLFOLDER" Name="BuiltInPropertiesExample" />
> 	</Directory>
> </Directory>
>
> I would like to install it under a different location according to
> property provided in the command line. for example :
> msiexec /i BuiltInPropertiesExample.msi
> **LOCATION=D:\MyProduct\BuiltInPropertiesExample**
> but if LOCATION property is not provided in the command, I would like to
> keep the current behavior: *C:\Program Files\BuiltInPropertiesExample*
> How do I make such installation ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/wixtoolset/discussions/7186>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACCKQ27Q7PGFMB7FFHVOEY3WVDRQTANCNFSM6AAAAAAUMHW55I>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Regards,
Fyodor Koryazhkin..

**Reply by idaamit** [reply]
I added : 
```
<SetDirectory Id="INSTALLFOLDER" Value="[LOCATION]" Sequence="execute" >
  <![CDATA[NOT Installed OR LOCATION]]>
</SetDirectory >
```

It doesn't  work :-(
what do you mean by  `after CostFinalize standard action`. how do I do it ?

**Comment by fyodorkor** [comment]
Please share your authoring (project)


On Tue, Jan 31, 2023 at 12:40 PM idaamit ***@***.***> wrote:

> I added :
>
> It doesn't work :-(
> what do you mean by after CostFinalize standard action. how do I do it ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/wixtoolset/discussions/7186#discussioncomment-4828415>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACCKQ27H4MCQ6EGOCIZJ7UTWVDTZDANCNFSM6AAAAAAUMHW55I>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Regards,
Fyodor Koryazhkin..

**Comment by idaamit** [comment]
It is working, I had problem in building the msi and it ran the old msi. Thank you

**Comment by robmen** [comment]
I may not fully understand your scenario, but the default behavior of the Windows Installer may work for you without needing to add a `SetDirectory`. You can change the `Id` of the `Directory` element to `LOCATION`, then set that via the command line will get you exactly the behavior you want (I believe).

I cover this topic in the Deployment Dojo episode [Last Minute Changes: Directories](https://www.youtube.com/live/y14Xd6VJJAw?feature=share).

**Reply by idaamit** [reply]
If I set the `Id` of the `Directory `element to `LOCATION` , What would happen if `LOCATION` is not provided in the command line ?
In case `LOCATION `was not provided I would like to install it under `ProgramFilesFolder`

**Reply by robmen** [reply]
I'm pretty sure that is covered in the episode.

**Reply by idaamit** [reply]
I have watched the episode and it is great!!

I have the following situation :

```
        <Directory Id="TARGETDIR" Name="SourceDir">
			<Directory Id="ProgramFilesFolder">
				<Directory Id="INSTALLFOLDER" Name="MyConpany">
                                     <Directory Id="subFolder" Name="MyProduct"/>
				</Directory>
			</Directory>
        </Directory>

```
I would like user have the ability to change the installation location but I would like my installation files to be always under _\MyCompany\MyProduct\_

for example : 
`msiexec /i BuiltInPropertiesExample.msi  INSTALLFOLDER=D:\AllPrograms\`

The product would be installed under : `D:\AllPrograms\MyProduct`,  but I would like it to be installed under `D:\AllPrograms\MyCompany\MyProduct`

In addition, what would happen if The provided folder would be invalid :  
1) unexisting drive
2) Invalid folder name , folder name greater than 255 characters
3) Invalid folder name , folder name with invalid characters  sach as :   ,<>

**Comment by fyodorkor** [comment]
Hi,
Try to change your authoring as following:

<Directory Id="TARGETDIR" Name="SourceDir">
			<Directory Id="ProgramFilesFolder">
				<Directory Id="MYCOMPANY" Name="MyConpany">
                                     <Directory Id="SUNFOLDER" Name="MyProduct">

                                       <Directory Id="INSTALLFOLDER" />

                                     <Directory/>
				</Directory>
			</Directory>
        </Directory>


On Thu, Feb 2, 2023 at 6:12 PM idaamit ***@***.***> wrote:

> I have watched the episode and it is great!!
>
> I have the following situation :
>
>         <Directory Id="TARGETDIR" Name="SourceDir">
> 			<Directory Id="ProgramFilesFolder">
> 				<Directory Id="INSTALLFOLDER" Name="MyConpany">
>                                      <Directory Id="sunFolder" Name="MyProduct"/>
> 				</Directory>
> 			</Directory>
>         </Directory>
>
>
> I would like user have the ability to change the installation location but
> I would like my installation files to be always under _\MyCompany\MyProduct_
>
> for example :
> msiexec /i BuiltInPropertiesExample.msi INSTALLFOLDER=D:\AllPrograms\
>
> The product would be installed under : D:\AllPrograms\MyProduct, but I
> would like it to be installed under D:\AllPrograms\MyCompany\MyProduct
>
> In addition, what would happen if The provided folder would be invalid :
>
>    1. unexisting drive
>    2. Invalid folder name , folder name greater than 255 characters
>    3. Invalid folder name , folder name with invalid characters sach as :
>    ,<>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/wixtoolset/discussions/7186#discussioncomment-4853361>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACCKQ22JZTMNWJKFWGQFQATWVPMF5ANCNFSM6AAAAAAUMHW55I>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Regards,
Fyodor Koryazhkin..

**Reply by idaamit** [reply]
@fyodorkor  I have tested your configuration (the above) , with the SetDirectory tag: 

```
    <SetDirectory Id="installationFolder" Value="[LOCATION]\MyConpany\MyProduct" Sequence="execute" >
      <![CDATA[LOCATION]]>
    </SetDirectory >

    
<Directory Id="TARGETDIR" Name="SourceDir">
	<Directory Id="ProgramFilesFolder">
            <Directory Id="MyConpany" Name="MyConpany">
                  <Directory Id="MyProduct" Name="MyProduct">
                           <Directory Id="installationFolder" />
                  </Directory>
        </Directory>
	</Directory>
</Directory>
```
If I do not provide LOCATION, the program is installed under /Program Files/MyConpany/MyProduct
If I provide LOCATION, 
`msiexec /i  BuiltInPropertiesExample.msi  INSTALLFOLDER=d:\FFFFF\`
the program is installed under d:\FFFFF\MyConpany\MyProduct

Thank you for your great assistance !

BTW , If I do not use the SetDirectory  , with the following code  : 		
```
<Directory Id="TARGETDIR" Name="SourceDir">
     <Directory Id="ProgramFilesFolder">
        <Directory Id="MyConpany" Name="MyConpany">
              <Directory Id="MyProduct" Name="MyProduct">
                    <Directory Id="INSTALLFOLDER " />
              </Directory>
        </Directory>
     </Directory>
</Directory>
```
The behaviour is : 
If I do not provide INSTALLFOLDER , the program is installed under /Program Files/MyConpany/MyProduct
If I provide INSTALLFOLDER , 
`msiexec /i  BuiltInPropertiesExample.msi  INSTALLFOLDER=d:\FFFFF\`
the program is installed under d:\FFFFF\  without _/MyConpany/MyProduct_

**Comment by idaamit** [comment]
BTW for  invalid path 
1) unexisting drive
2) Invalid folder name , folder name greater than 255 characters
3) Invalid folder name , folder name with invalid characters sach as : ,<>

We get the following error messages : 
1) Invalid Drive 
2) The portion of the folder path is invalid, it is either empty or exceeds the length allowed by the system
3) The folder path contains an invalid character 
or 
The filename, directory name, or volume label syntax is incorrect
 
looks good
